enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+to spread the Machines across multiple failure domains.
+
+#### Creation of new Machines
+
+When a Machine API MachineSet creates a new Machine, the synchronisation controller will create the equivalent Cluster API Machine based on the method described in [Machine Synchronisation](#machine-synchronisation).
+
+When a Cluster API MachineSet creates a new Machine, the synchronisation controller will create it’s Machine API equivalent if, and only if, the Cluster API MachineSet has a Machine API equivalent.
+This will allow users to switch a MachineSet back to Machine API at a later date if they wish to do so.
+
+When the Cluster API MachineSet has been created independently of Machine API, new Machines will not be created in Machine API to reflect the Machines owned by the Cluster API MachineSet.
+
+#### Deletion of Machines
+
+To enable the synchronisation of deletion, the synchronisation controller will use its own Finalizer on resources that have Machine API and Cluster API equivalents.
+
+This means that for a Machine, we expect the authoritative controller to perform the actual deletion work within the infrastructure provider.
+Once it has completed this effort, it will remove its own Finalizer.
+
+As soon as the synchronisation controller notices the Machine has been deleted, it should ensure that the Machine API/Cluster API equivalent (if present) is also deleted.
+It should then wait until the authoritative Machine controller removes its finalizer before removing both its own, and the Machine API/Cluster API Machine finalizer if present.
+In scenarios where both Machine API and Cluster API have been authoritative, it is expected that both Machine controllers will have added their own Finalizers, therefore we expect the synchronisation controller should be able to handle this.
+
+It is also feasible that a customer may want to remove the Machine API resources after they have migrated to Cluster API and no longer require the Machine API synchronisation.
+To allow for this, if a non-authoritative Machine API parent[^8] resource is deleted, the deletion event will not be synchronised to the Cluster API equivalent and the synchronisation controller will ensure finalizers are removed as appropriate.
+In this scenario, the Cluster API equivalent will continue to operate on its own.
+To ensure users are aware of this situation, webhooks will be introduced to warn users that the deletion will not be synchronised.
+
+[^8]: To allow continued operation of MachineSets, ControlPlaneMachineSets and MachineHealthChecks, if a Machine API Machine that is owned is deleted,
+then the deletion event is synchronised to Cluster API.
+This will allow for a MachineSet/ControlPlaneMachineSet to own mixed instances (i.e. those managed by both Machine API and Cluster API), and continue to support higher level remediation automation.
```

Comment:
> WHO generates CAPI Machines??

The intention is for the syncer to do this. It will have logic for mirroring machines anyway, as described, so it will notice the new machines created by the MAPI machineset and then treat them as if they were any other machine, mirroring them. We cannot have the CAPI machineset do the mirroring since it would not guarantee that the names of the machines in both namespaces match, in which case you would end up with more machines than intended

> If at this point the admin flips the CAPI MachineSet to authoritative... hopefully this is a no-op, but could result in the CAPI Machines getting modified

This should be a no-op yes, the machineset controllers don't update machines once created, they just check the replica counts. So, as long as we get the logic correct for how the machineset identifies machines that it owns, we should be ok

>  so, the syncer will push changes from the CAPI-authoritative subset back to their MAPI mirrors... but will also try to undo the changes from the MAPI-authoritative Machines back to their CAPI mirrors, resulting in thrash.

Based on some other discussion, we are making some adjustments to how the sync handover will work, so that the sync controller is responsible for ensuring the sync is complete before making the switch of authority. This should prevent thrashing

> To mitigate this, I think (both) MachineSet controllers need special logic to detect when the Machines they would be reconciling are marked as non-authoritative, and ignore them. This may be implied by words elsewhere, but seems worth an explicit callout (and test cases!).

The only real logic that the machineset switch over is handling is the way the Machines are created. By default, a MAPI machineset will create MAPI authoritative machines, however, you could update the MAPI machineset spec to create CAPI authoritative machines, which would test the conversion and creation of CAPI machines. The opposite is true for a CAPI machineset, except we won't start mirroring CAPI to MAPI if there's no MAPI equivalent machineset.

Testing the Machineset behaviours is the real point in the Machineset authority. The behaviour between them is near identical, except in how they handle replacing machines. If you delete a Machine in MAPI, it has to be removed from the cluster before it is replaced by the MachineSet. In CAPI, a new Machine is created as soon as the deletion timestamp is present. To my knowledge, this is the only interesting difference that switch the MachineSet to authoritative really achieves